src/App.tsx

```diff
@@ -1,24 +1,198 @@
+    case FilterTodos.completed:
+      return todos.filter((todo) => todo.completed);
+    default:
+      return todos;
+  }
+};
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [query, setQuery] = useState('');
+  const [errorMessage, setErrorMessage] = useState('');
+  const [filter, setFilter] = useState<FilterTodos | ''>('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isTodoLoading, setIsTodoLoading] = useState<Todo | null>(null);
+  const [isInputDisabled, setIsInputDisabled] = useState(false);
+
+  const errorTimerId = useRef(0);
+
+  const showErrorMessage = () => {
+    if (errorTimerId.current) {
+      clearTimeout(errorTimerId.current);
+    }
+
+    errorTimerId.current = window.setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+  };
+
+  useEffect(() => {
+    showErrorMessage();
+  }, [errorMessage]);
+
+  useEffect(() => {
+    setErrorMessage('');
+
+    dataOperations
+      .getTodos(USER_ID)
+      .then(setTodos)
+      .catch((error) => {
+        setErrorMessage('Unable to load todos');
+        setTimeout(() => {
+          setErrorMessage('');
+        }, 3000);
+        throw error;
+      });
+  }, []);
+
+  const filteredTodos = useMemo(() => {
+    return filterTodos(todos, filter);
+  }, [todos, filter]);
+
+  const handleFilter = useCallback((newFilter: FilterTodos) => {
+    setFilter(newFilter);
+  }, [setFilter]);
+
+  const count = todos.reduce((acc, todo) => (todo.completed ? acc : acc + 1), 0);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const addTodo = ({ userId, title, completed }: Todo) => {
+    setErrorMessage('');
+
+    const temporaryTodo = {
+      userId,
+      title,
+      completed,
+      id: 0,
+    };
+
+    setTempTodo(temporaryTodo);
+    setIsTodoLoading(temporaryTodo);
+    setIsInputDisabled(true);
+
+    return dataOperations
+      .addTodoToServer({ userId, title, completed })
+      .then((newTodo) => {
+        setTodos((currentTodos) => [...currentTodos, newTodo]);
+      })
+      .catch((error) => {
+        setTodos((prev) => prev.filter((todo) => todo.id !== temporaryTodo.id));
+        setErrorMessage('Unable to add a todo');
+        setTimeout(() => {
+          setErrorMessage('');
+        }, 3000);
+        throw error;
+      })
+      .finally(() => {
+        setTempTodo(null);
+        setIsInputDisabled(false);
+        setIsTodoLoading(null);
+      });
+  };
+
+  const deleteTodo = (todoId: number) => {
+    setErrorMessage('');
+
+    return dataOperations
+      .deleteTodoOnServer(todoId)
+      .then(() => {
+        setTodos((currentTodo) => currentTodo.filter((todo) => todo.id !== todoId));
+      })
+      .catch((error) => {
+        setTodos(todos);
+        setErrorMessage('Unable to delete a todo');
+        setTimeout(() => {
+          setErrorMessage('');
+        }, 3000);
+        throw error;
```

Comment:
```suggestion

```

Reply:
if I don't raise the error further I won't be able to get it when the request to the server failed